hw/misc/hermes.c

```diff
@@ -398,21 +399,52 @@ static inline void hermes_bar_warn_unimplemented(unsigned bar, hwaddr addr)
                 addr);
 }
 
-static uint64_t hermes_exec(void *opaque, int engine)
+static void hermes_exec(void *opaque, int engine, int *response)
 {
+    HermesState *hermes = opaque;
     void *base = (uint32_t *) &((uint8_t *) &bar0_init)[0x1000 + engine * 48];
     int command = *(int *) base;
 
     if (command != 0x80) {
         hermes_bar_warn_invalid_command(command, engine);
-        return 0x1; /* TODO: Is there an error code for invalid command? */
+        return; /* TODO: Is there an error code for invalid command? */
     }
 
     int program_slot = *(int *)(base + 8);
     int data_slot = *(int *)(base + 9);
 
-    /* Run program slot */
-    return 0x0;
+    void *bar4_base = memory_region_get_ram_ptr(&hermes->bar4_mem_reg);
+    void *program_base = bar4_base + bar0_init.ehpsoff +
+                         (program_slot * bar0_init.ehpssze);
+    void *data_base = bar4_base + bar0_init.ehdsoff +
+                      (data_slot * bar0_init.ehdssze);
+
+    char *errmsg;
+    int32_t rv;
+    uint64_t ret;
+    bool elf;
+
+    /* Check magic number (first 4 bytes), to see if program is in ELF format */
+    elf = bar0_init.ehpssze >= 4 && !memcmp(program_base, ELFMAG, 4);
+    if (elf) {
+        rv = ubpf_load_elf(hermes->engines[engine], program_base,
+                           bar0_init.ehpssze, &errmsg);
+    } else {
+        rv = ubpf_load(hermes->engines[engine], program_base,
+                       bar0_init.ehpssze, &errmsg);
+    }
+
+    if (rv < 0) {
+        response[32] = 0x05;
```

Comment:
0x05 should be a `#define` or part of an enum with all other possible status codes